pkg/query-service/auth/auth.go

```diff
@@ -207,65 +209,121 @@ type RegisterRequest struct {
+func RegisterFirstUser(ctx context.Context, req *RegisterRequest) (*model.User, *model.ApiError) {
 
-	zap.S().Debugf("Got a register request for email: %v\n", req.Email)
+	if req.Email == "" {
+		return nil, model.BadRequest(model.ErrEmailRequired{})
+	}
 
-	// TODO(Ahsan): We should optimize it, shouldn't make an extra DB call everytime to know if
-	// this is the first register request.
-	users, apiErr := dao.DB().GetUsers(ctx)
+	if req.Password == "" {
+		return nil, model.BadRequest(model.ErrPasswordRequired{})
+	}
+
+	groupName := constants.AdminGroup
+
+	org, apierr := dao.DB().CreateOrg(ctx,
+		&model.Organization{Name: req.OrgName})
+	if apierr != nil {
+		zap.S().Debugf("CreateOrg failed, err: %v\n", zap.Error(apierr.ToError()))
+		return nil, apierr
+	}
+
+	group, apiErr := dao.DB().GetGroupByName(ctx, groupName)
 	if apiErr != nil {
-		zap.S().Debugf("GetUser failed, err: %v\n", apiErr.Err)
-		return apiErr
+		zap.S().Debugf("GetGroupByName failed, err: %v\n", apiErr.Err)
+		return nil, apiErr
 	}
 
-	var groupName, orgId string
+	var hash string
+	var err error
 
-	// If there are no user, then this first user is granted Admin role. Also, an org is created
-	// based on the request. Any other user can't use any other org name, if they do then
-	// registration will fail because of foreign key violation while create user.
-	// TODO(Ahsan): We need to re-work this logic for the case of multi-tenant system.
-	if len(users) == 0 {
-		org, apiErr := dao.DB().CreateOrg(ctx, &model.Organization{Name: req.OrgName})
-		if apiErr != nil {
-			zap.S().Debugf("CreateOrg failed, err: %v\n", apiErr.Err)
-			return apiErr
-		}
-		groupName = constants.AdminGroup
-		orgId = org.Id
+	hash, err = passwordHash(req.Password)
+	if err != nil {
+		zap.S().Errorf("failed to generate password hash when registering a user", zap.Error(err))
+		return nil, model.InternalError(model.ErrSignupFailed{})
 	}
 
-	if len(users) > 0 {
-		inv, err := validateInvite(ctx, req)
-		if err != nil {
-			return &model.ApiError{Err: err, Typ: model.ErrorUnauthorized}
-		}
-		org, apiErr := dao.DB().GetOrgByName(ctx, req.OrgName)
-		if apiErr != nil {
-			zap.S().Debugf("GetOrgByName failed, err: %v\n", apiErr.Err)
-			return apiErr
-		}
+	user := &model.User{
+		Id:                 uuid.NewString(),
+		Name:               req.Name,
+		Email:              req.Email,
+		Password:           hash,
+		CreatedAt:          time.Now().Unix(),
+		ProfilePirctureURL: "", // Currently unused
+		GroupId:            group.Id,
+		OrgId:              org.Id,
+	}
 
-		groupName = inv.Role
-		if org != nil {
-			orgId = org.Id
-		}
+	return dao.DB().CreateUser(ctx, user)
+}
+
+// RegisterInvitedUser handles registering a invited user
+func RegisterInvitedUser(ctx context.Context, req *RegisterRequest, nopassword bool) (*model.User, *model.ApiError) {
+
+	if req.InviteToken == "" {
+		return nil, model.BadRequest(fmt.Errorf("invite token is required"))
 	}
```

Comment:
Same here. can this be done at request data parsing unless it requires fetching some data from DB.

Reply:
IMO the methods - especially that contain business logic -  should clearly validate inputs they expect and throw error if they don't have it.  

The pattern of depending on HTTP handlers to validate may not be best, as we may re-use such methods in CLI or other places.  This will also reduce duplication of maintaining validations at both base and ee api handler.